docs/upgrading/v6-data.md

```diff
@@ -0,0 +1,216 @@
+
+- Route-level data APIs such as `loader`, `action`, `shouldRevalidate`, `handle`, and `lazy`
+- In-component data hooks such as `useLoaderData`, `useActionData`, `useFetcher`, `useMatches`, `useNavigation`, etc.
+- Error-handling APIs such as `route.errorElement`, `route.ErrorBoundary`, and `useRouteError`
+
+The rest of the APIs that existed prior to v6.4.0 are still usable in _both_ `BrowserRouter` and `RouterProvider` apps. These include common hooks/components such as `useNavigate`, `useLocation`, `useParams`, `<Link>`, `<Outlet />`, etc.
+
+## Migrating
+
+We built the new `<RouterProvider>` component such that it would enable the new Data APIs on routes defined at the root router, while not excluding descendant `<Routes>` trees so commonly used in `BrowserRouter` apps. This was explicitly to allow incremental migration from one to the other. Let's take a look at how we would do this.
+
+### Current Application
+
+Let's assume we've got a current application with 2 descendant route trees, and assume these routes are all doing in-component data fetching, and rendering their own loading and error states.
+
+```tsx
+import * as React from "react";
+import {
+  BrowserRouter,
+  Link,
+  Route,
+  Routes,
+} from "react-router-dom";
+
+export default function App() {
+  return (
+    <BrowserRouter>
+      <Routes>
+        <Route path="/" element={<Home />} />
+        <Route path="/blog/*" element={<BlogApp />} />
+        <Route path="/users/*" element={<UserApp />} />
+      </Routes>
+    </BrowserRouter>
+  );
+}
+
+function Home() {
+  return (
+    <>
+      <h1>Welcome!</h1>
+      <p>
+        Check out the <Link to="/blog">blog</Link> or the{" "}
+        <Link to="users">users</Link> section
+      </p>
+    </>
+  );
+}
+
+function BlogApp() {
+  return (
+    <Routes>
+      <Route index element={<h1>Blog Index</h1>} />
+      <Route path="posts" element={<h1>Blog Posts</h1>} />
+    </Routes>
+  );
+}
+
+function UserApp() {
+  return (
+    <Routes>
+      <Route index element={<h1>Users Index</h1>} />
+    </Routes>
+  );
+}
+```
+
+### Add RouterProvider with a root splat route
+
+We can render this application inside a `RouterProvider` with only a few small changes:
+
+1. Change your current `App` component to `Root`
+2. Remove the `<BrowserRouter>` component
+3. Create a data router singleton with a splat route for the `Root` element
+4. Add a new `App` component rendering a `<RouterProvider>`
+
+```tsx lines=[10-13,15-18,20-21,22-23]
+import * as React from "react";
+import {
+  createBrowserRouter,
+  Link,
+  Route,
+  RouterProvider,
+  Routes,
+} from "react-router-dom";
+
+// 3️⃣ Router singleton created
+const router = createBrowserRouter([
+  { path: "*", Component: Root },
+]);
+
+// 4️⃣ RouterProvider added
+export default function App() {
+  return <RouterProvider router={router} />;
+}
+
+// 1️⃣ Changed from App to Root
+function Root() {
```

Comment:
ig moving this to a separate component rather than keeping everything in `App` makes the visible diff simpler?

Reply:
ah, it makes moving from the JSX Route to the JSON Route more apparent as you lift it up..